### PR TITLE
feat: Update supply borrow e2e tests (SC-80)

### PR DESCRIPTION
**Pull request by lucas-manuel**
This PR does the following:

1. Tests all permutations of collateral assets and borrow assets
2. Adds a test to check the edge condition of LTV borrowing, causing a failure and then a success
3. Adds a test to borrow, repay, and withdraw with variable rates
4. Adds a test to borrow, repay, and withdraw with stable rates (for upsteam PR)
5. Adds warping to tests
6. Ensures interest is accruing against position, preventing full withdrawal of collateral until additional funds are added
7. Adds full reserve state assertions to all E2E tests
8. Adds all token balance assertions to supply, withdraw, borrow, repay functions
9. Formats console log output for decimals
10. Adds naive calculation for interest accrual, asserts it against warp
11. Updated deposit to supply

### CODE REVIEW

**Comment by hexonaut** [comment]
Going to wait until you're back to review this together as this is not compiling atm and is still WIP tagged.

File: src/ProtocolV3TestBase.sol [hunk 1]
```diff
@@ -127,53 +140,81 @@ contract ProtocolV3TestBase is CommonTestBase {
...
   }
 
   /**
    * Reserves that are frozen or not active should not be included in e2e test suite
    */
-  function _includeInE2e(ReserveConfig memory config) internal pure returns (bool) {
-    return !config.isFrozen && config.isActive && !config.isPaused;
+  // TODO: Fix LDO issue
+  function _includeBorrowAssetInE2e(ReserveConfig memory config) internal pure returns (bool) {
+    return !config.isFrozen && config.isActive && !config.isPaused && config.borrowingEnabled && config.underlying != 0x5A98FcBEA516Cf06857215779Fd812CA3beF1B32;
```
Comment (lucas-manuel) [inline src/ProtocolV3TestBase.sol]: Couldn't figure out a way around this, its an issue with `deal` looks like

File: src/ProtocolV3TestBase.sol [hunk 2]
```diff
@@ -127,53 +138,83 @@ contract ProtocolV3TestBase is CommonTestBase {
...
    * Reserves that are frozen or not active should not be included in e2e test suite
    */
-  function _includeInE2e(ReserveConfig memory config) internal pure returns (bool) {
-    return !config.isFrozen && config.isActive && !config.isPaused;
+  function _includeBorrowAssetInE2e(ReserveConfig memory config) internal pure returns (bool) {
+    return !config.isFrozen && config.isActive && !config.isPaused && config.borrowingEnabled;
+  }
+
+  function _includeCollateralAssetInE2e(ReserveConfig memory config) internal pure returns (bool) {
+    return !config.isFrozen && config.usageAsCollateralEnabled;
```
Comment (hexonaut) [inline src/ProtocolV3TestBase.sol]: Should also be testing ` && config.isActive && !config.isPaused` ?

File: src/ProtocolV3TestBase.sol [hunk 3]
```diff
@@ -184,68 +225,202 @@ contract ProtocolV3TestBase is CommonTestBase {
...
+  ) internal view returns (uint256) {
+    IPoolAddressesProvider addressesProvider = IPoolAddressesProvider(pool.ADDRESSES_PROVIDER());
+    IAaveOracle oracle = IAaveOracle(addressesProvider.getPriceOracle());
+    return collateralAmount
+      * oracle.getAssetPrice(collateralConfig.underlying)
+      * collateralConfig.ltv
+      * (10 ** borrowConfig.decimals)
+      / oracle.getAssetPrice(borrowConfig.underlying)
+      / (10 ** collateralConfig.decimals)
+      / 10_000;
```
Comment (hexonaut) [inline src/ProtocolV3TestBase.sol]: This is a bps value so I usually write it like `100_00` to signify 100%.

File: src/ProtocolV3TestBase.sol [hunk 4]
```diff
@@ -184,68 +225,202 @@ contract ProtocolV3TestBase is CommonTestBase {
...
+    _assertReserveChange(beforeReserve, afterReserve, int256(amount), 60 seconds);
+
+    // Step 4: Try to withdraw all collateral, demonstrate it's not possible without paying back
+    //         accrued debt
+
+    uint256 totalCollateral = IERC20(collateralConfig.aToken).balanceOf(borrower);
+    uint256 remainingDebt   = IERC20(debtToken).balanceOf(borrower);
+
+    // Handle edge case for DAI LTV at 0.01% causing rounding errors here, preventing failure.
+    if (collateralConfig.ltv > 1) {
```
Comment (hexonaut) [inline src/ProtocolV3TestBase.sol]: Remove the mention of DAI and the ltv should not be specific to the DAI market. I would actually set a lower bound on selecting whether something is "good collateral" in the `_includeCollateralAssetInE2e()` function. That function should skip the DAI market in Spark because it is effectively not collateral. Something like `ltv > 5%` or something to ensure reasonable numbers.

File: src/ProtocolV3TestBase.sol [hunk 5]
```diff
@@ -184,68 +225,202 @@ contract ProtocolV3TestBase is CommonTestBase {
...
+
+    uint256 expectedInterest;
+    for (uint256 i; i < timeSinceLastUpdate; i++) {
+      expectedInterest +=
+        uint256(beforeReserve.variableBorrowIndex)
+        * uint256(beforeReserve.currentVariableBorrowRate)
+        * 1 seconds
+        / 365 days
+        / 1e27;
     }
```
Comment (hexonaut) [inline src/ProtocolV3TestBase.sol:381]: This should probably be optimized, but is fine for now.

File: src/ProtocolV3TestBase.sol [hunk 6]
```diff
@@ -184,68 +225,202 @@ contract ProtocolV3TestBase is CommonTestBase {
...
+  function _supply(
     ReserveConfig memory config,
     IPool pool,
     address user,
     uint256 amount
   ) internal {
     require(!config.isFrozen, 'DEPOSIT(): FROZEN_RESERVE');
-    require(config.isActive, 'DEPOSIT(): INACTIVE_RESERVE');
+    require( config.isActive, 'DEPOSIT(): INACTIVE_RESERVE');
     require(!config.isPaused, 'DEPOSIT(): PAUSED_RESERVE');
```
Comment (hexonaut) [inline src/ProtocolV3TestBase.sol]: Revert error should be changed to `SUPPLY(): XXX`.

**Review by hexonaut — CHANGES_REQUESTED** [review]
Mostly minor stuff. Good to merge after these things addressed.

File: src/ProtocolV3TestBase.sol (hunk 2 above)
Comment (lucas-manuel) [reply src/ProtocolV3TestBase.sol]: Checked ValidationLogic and yes should be added, added

File: src/ProtocolV3TestBase.sol (hunk 3 above)
Comment (lucas-manuel) [reply src/ProtocolV3TestBase.sol]: Updated

File: src/ProtocolV3TestBase.sol (hunk 4 above)
Comment (lucas-manuel) [reply src/ProtocolV3TestBase.sol]: Updated, slight correction on what we spoke on:

This function prevents the check of the full collateral amount not being able to be withdrawn because of interest generated on the borrow asset. In the case of low LTV collaterals (e.g., DAI) the amount of asset that can be borrowed is so low that the interest accrued doesn't change the LTV. This condition is still good though because it will test DAI as any other collateral type.

File: src/ProtocolV3TestBase.sol (hunk 6 above)
Comment (lucas-manuel) [reply src/ProtocolV3TestBase.sol]: Updated

**Review by hexonaut — APPROVED** [review]
Good to merge.
